# pandas Fundamentals
McKinney Ch 5 (basics) + Ch 7 (data cleaning). Ch 6 has no exercises

In [1]:
import numpy as np
import pandas as pd

---
## Chapter 5 — pandas Basics

### Series

In [2]:
obj = pd.Series([4, 7, -5, 3])
obj
obj.array

<NumpyExtensionArray>
[4, 7, -5, 3]
Length: 4, dtype: int64

In [3]:
obj.index

RangeIndex(start=0, stop=4, step=1)

In [4]:
obj2 = pd.Series([4, 7, -5, 3], index=['d', 'b', 'a', 'c'])
obj2

d    4
b    7
a   -5
c    3
dtype: int64

In [5]:
obj2['a']
obj2[['c', 'a', 'd']]

c    3
a   -5
d    4
dtype: int64

In [6]:
obj2[obj2 > 0] #only keep positive values
obj2 * 2

d     8
b    14
a   -10
c     6
dtype: int64

In [7]:
np.exp(obj2)

d      54.598150
b    1096.633158
a       0.006738
c      20.085537
dtype: float64

A series maps index values (left column) to actual data values
- similar to a dictionary
- dict = unformatted, array style listing

In [8]:
"b" in obj2
"e" in obj2

False

In [9]:
sdata = {"Ohio": 35000, "Texas": 71000, "Oregon": 16000, "Utah": 5000}
obj3 = pd.Series(sdata)
obj3

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

In [10]:
obj3.to_dict()

{'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}

In [11]:
states = ["California", "Ohio", "Oregon", "Texas"] # California is not in sdata, so it will be NaN
obj4 = pd.Series(sdata, index=states)
obj4

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64

In [12]:
pd.isna(obj4) #check if values are NaN
pd.notna(obj4)

California    False
Ohio           True
Oregon         True
Texas          True
dtype: bool

In [13]:
obj3 + obj4 #arithmethic between different indexes. NaN is like undefined

California         NaN
Ohio           70000.0
Oregon         32000.0
Texas         142000.0
Utah               NaN
dtype: float64

In [14]:
obj4.name = "population" #name the entire Series
obj4.index.name = "state"
obj4

state
California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
Name: population, dtype: float64

In [15]:
obj
obj.index = ['Bob', 'Steve', 'Jeff', 'Ryan'] #adding an index
obj

Bob      4
Steve    7
Jeff    -5
Ryan     3
dtype: int64

### DataFrame
dictionary of Series' with the same index

In [16]:
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003], 
        "pop": [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
frame = pd.DataFrame(data)
frame

,state,year,pop
0,Ohio,2000,1.5
1,Ohio,2001,1.7
2,Ohio,2002,3.6
3,Nevada,2001,2.4
4,Nevada,2002,2.9
5,Nevada,2003,3.2


In [17]:
frame.head() #show only first 5 rows
frame.tail(3) #show only last 3 rows

,state,year,pop
3,Nevada,2001,2.4
4,Nevada,2002,2.9
5,Nevada,2003,3.2


In [18]:
pd.DataFrame(data, columns=['year', 'state', 'pop']) #specify column order

,year,state,pop
0,2000,Ohio,1.5
1,2001,Ohio,1.7
2,2002,Ohio,3.6
3,2001,Nevada,2.4
4,2002,Nevada,2.9
5,2003,Nevada,3.2


In [19]:
frame2 = pd.DataFrame(data, columns=['year', 'state', 'pop', 'debt'])
frame2

,year,state,pop,debt
0,2000,Ohio,1.5,NaN
1,2001,Ohio,1.7,NaN
2,2002,Ohio,3.6,NaN
3,2001,Nevada,2.4,NaN
4,2002,Nevada,2.9,NaN
5,2003,Nevada,3.2,NaN


In [20]:
frame2.columns

Index(['year', 'state', 'pop', 'debt'], dtype='object')

In [21]:
frame2["state"] #returns Series of a single column
frame2.year #works the same way

0    2000
1    2001
2    2002
3    2001
4    2002
5    2003
Name: year, dtype: int64

### .loc vs .iloc
special attributes to select rows and columns from a DataFrame

In [22]:
frame2.loc[1]

year     2001
state    Ohio
pop       1.7
debt      NaN
Name: 1, dtype: object

In [23]:
frame2.iloc[2]

year     2002
state    Ohio
pop       3.6
debt      NaN
Name: 2, dtype: object

iloc = uses integer
loc = uses labels (actual index values and column names) 

### Index alignment in arithmetic

In [24]:
frame2["debt"] = 16.5
frame2

,year,state,pop,debt
0,2000,Ohio,1.5,16.5
1,2001,Ohio,1.7,16.5
2,2002,Ohio,3.6,16.5
3,2001,Nevada,2.4,16.5
4,2002,Nevada,2.9,16.5
5,2003,Nevada,3.2,16.5


In [25]:
frame2["debt"] = np.arange (6.)
frame2

,year,state,pop,debt
0,2000,Ohio,1.5,0.0
1,2001,Ohio,1.7,1.0
2,2002,Ohio,3.6,2.0
3,2001,Nevada,2.4,3.0
4,2002,Nevada,2.9,4.0
5,2003,Nevada,3.2,5.0


In [26]:
val = pd.Series([-1.2, -1.5, -1.7], index=["two", "four", "five"])
frame2["debt"] = val
frame2

,year,state,pop,debt
0,2000,Ohio,1.5,NaN
1,2001,Ohio,1.7,NaN
2,2002,Ohio,3.6,NaN
3,2001,Nevada,2.4,NaN
4,2002,Nevada,2.9,NaN
5,2003,Nevada,3.2,NaN


In [27]:
frame2["eastern"] = frame2["state"] == "Ohio"
frame2
del frame2["eastern"]
frame2.columns

Index(['year', 'state', 'pop', 'debt'], dtype='object')

In [28]:
populations = {"Ohio": {2000: 1.5, 2001: 1.7, 2002: 3.6},
               "Nevada": {2001: 2.4, 2002: 2.9}}
frame3 = pd.DataFrame(populations)
frame3

,Ohio,Nevada
2000,1.5,NaN
2001,1.7,2.4
2002,3.6,2.9


In [29]:
frame3.T

,2000,2001,2002
Ohio,1.5,1.7,3.6
Nevada,NaN,2.4,2.9


In [30]:
pd.DataFrame(populations, index=[2001, 2002, 2003])

,Ohio,Nevada
2001,1.7,2.4
2002,3.6,2.9
2003,NaN,NaN


In [31]:
pdata = {"Ohio": frame3["Ohio"][:-1], 
         "Nevada": frame3["Nevada"][:2]}
pd.DataFrame(pdata)

,Ohio,Nevada
2000,1.5,NaN
2001,1.7,2.4


*slice notation*

[start:stop]
stop number is always excluded

In [32]:
frame3.index.name = "year"
frame3.columns.name = "state"
frame3 #unlike Series, DataFrame doesn't have a name attribute

state,Ohio,Nevada
year,,
2000,1.5,NaN
2001,1.7,2.4
2002,3.6,2.9


In [33]:
frame3.to_numpy()
frame2.to_numpy()

array([[2000, 'Ohio', 1.5, nan],
       [2001, 'Ohio', 1.7, nan],
       [2002, 'Ohio', 3.6, nan],
       [2001, 'Nevada', 2.4, nan],
       [2002, 'Nevada', 2.9, nan],
       [2003, 'Nevada', 3.2, nan]], dtype=object)

In [34]:
obj = pd.Series(np.arange(3), index=['a', 'b', 'c'])
index = obj.index
index
index[1:] 

Index(['b', 'c'], dtype='object')

In [35]:
labels = pd.Index(np.arange(3))
labels

Index([0, 1, 2], dtype='int64')

In [36]:
obj2 = pd.Series([1.5, -2.5, 0], index=labels)
obj2
obj2.index is labels

True

Index are kinda like arrays

In [37]:
frame3
frame3.columns
"Ohio" in frame3.columns #is Ohio in the columns of frame3
2003 in frame3.index #is 2003 in the index of frame3 (no)

False

In [38]:
pd.Index(["foo", "foo", "bar", "bar"])

Index(['foo', 'foo', 'bar', 'bar'], dtype='object')

### Reindexing

In [39]:
obj = pd.Series([4.5, 7.2, -5.3, 3.6], index=['d', 'b', 'a', 'c'])
obj

d    4.5
b    7.2
a   -5.3
c    3.6
dtype: float64

In [40]:
obj2 = obj.reindex(['a', 'b', 'c', 'd', 'e']) #reindexing to a new set of indexes
obj2

a   -5.3
b    7.2
c    3.6
d    4.5
e    NaN
dtype: float64

In [41]:
obj3 = pd.Series(['blue', 'purple', 'yellow'], index=[0, 2, 4])
obj3

0      blue
2    purple
4    yellow
dtype: object

In [42]:
obj3.reindex(np.arange(6), method="ffill") #forward fill method for reindexing

0      blue
1      blue
2    purple
3    purple
4    yellow
5    yellow
dtype: object

In [43]:
frame = pd.DataFrame(np.arange(9).reshape((3, 3)), 
                     index=['a', 'c', 'd'], 
                     columns=['Ohio', 'Texas', 'California'])
frame

,Ohio,Texas,California
a,0,1,2
c,3,4,5
d,6,7,8


In [44]:
frame2 = frame.reindex(index=['a', 'b', 'c', 'd'])
frame2

,Ohio,Texas,California
a,0.0,1.0,2.0
b,NaN,NaN,NaN
c,3.0,4.0,5.0
d,6.0,7.0,8.0


In [45]:
states = ["Texas", "Utah", "California"]
frame.reindex(columns=states) #reindex with a keyword

,Texas,Utah,California
a,1,NaN,2
c,4,NaN,5
d,7,NaN,8


In [46]:
frame.reindex(states, axis="columns") #same way to do ^^ with axis keyword instead

,Texas,Utah,California
a,1,NaN,2
c,4,NaN,5
d,7,NaN,8


In [47]:
frame.loc[["a", "d", "c"], ["California", "Texas"]] #reindexing with loc, changing both rows and columns at once

,California,Texas
a,2,1
d,8,7
c,5,4


frame.loc only works if all the new index labels already exist
- reindex adds NaN for missing labels

### Dropping Entries from an Axis

In [48]:
obj = pd.Series(np.arange(5.), index=['a', 'b', 'c', 'd', 'e'])
obj

a    0.0
b    1.0
c    2.0
d    3.0
e    4.0
dtype: float64

In [49]:
new_obj = obj.drop("c") #drop an index (row c) from the Series
new_obj
obj.drop(["d", "c"])

a    0.0
b    1.0
e    4.0
dtype: float64

In [50]:
data = pd.DataFrame(np.arange(16).reshape((4, 4)),
                    index=['Ohio', 'Colorado', 'Utah', 'New York'],
                    columns=['one', 'two', 'three', 'four'])
data

,one,two,three,four
Ohio,0,1,2,3
Colorado,4,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


In [51]:
data.drop(columns=["two"])

,one,three,four
Ohio,0,2,3
Colorado,4,6,7
Utah,8,10,11
New York,12,14,15


In [52]:
data.drop(index=["Colorado", "Ohio"])

,one,two,three,four
Utah,8,9,10,11
New York,12,13,14,15


In [53]:
data.drop("two", axis=1) #dropping columns with axis keyword instead of columns keyword like NumPy

,one,three,four
Ohio,0,2,3
Colorado,4,6,7
Utah,8,10,11
New York,12,14,15


In [54]:
data.drop(["two", "four"], axis="columns") #drop columns two and four

,one,three
Ohio,0,2
Colorado,4,6
Utah,8,10
New York,12,14


### Indexing, Selection, and Filtering

In [55]:
obj = pd.Series(np.arange(4.), index=['a', 'b', 'c', 'd'])
obj

a    0.0
b    1.0
c    2.0
d    3.0
dtype: float64

In [56]:
obj.loc['b'] #loc looks for label first column, not integer position. 
obj.iloc [1] #iloc looks for integer position, not label.

1.0

In [57]:
obj[2:4] #slicing

c    2.0
d    3.0
dtype: float64

In [58]:
obj[["b", "a", "d"]]

b    1.0
a    0.0
d    3.0
dtype: float64

In [59]:
obj [[1, 3]]

/var/folders/9p/c9c7v_v53bd91dc97zsmsm740000gn/T/ipykernel_30191/647580651.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  obj [[1, 3]]


b    1.0
d    3.0
dtype: float64

In [60]:
obj[obj < 2] #boolean indexing

a    0.0
b    1.0
dtype: float64

In [61]:
obj.loc[["b", "a", "d"]] #preferred to label with loc

b    1.0
a    0.0
d    3.0
dtype: float64

In [62]:
obj1 = pd.Series([1, 2, 3], index=[2, 0, 1])
obj2 = pd.Series([1, 2, 3], index=["a", "b", "c"])
obj1

2    1
0    2
1    3
dtype: int64

In [63]:
obj1[[0, 1, 2]] #sorting out the order of the index, not the position of the index.

0    2
1    3
2    1
dtype: int64

In [64]:
obj2 [[0, 1, 2]] #sorting out doesn't change order for this set

/var/folders/9p/c9c7v_v53bd91dc97zsmsm740000gn/T/ipykernel_30191/2163431866.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  obj2 [[0, 1, 2]] #sorting out doesn't change order for this set


a    1
b    2
c    3
dtype: int64

In [65]:
obj1. iloc[[0, 1, 2]] #sorting out the position of the index, not the order of the index. iloc is always integer position based.
obj2.iloc[[0, 1, 2]] #using iloc instead to reorder

a    1
b    2
c    3
dtype: int64

In [66]:
obj2.loc["b":"c"] = 5
obj2

a    1
b    5
c    5
dtype: int64

In [67]:
data = pd.DataFrame(np.arange(16).reshape((4, 4)),
                    index=['Ohio', 'Colorado', 'Utah', 'New York'],
                    columns=['one', 'two', 'three', 'four'])
data

,one,two,three,four
Ohio,0,1,2,3
Colorado,4,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


In [68]:
data["two"]
data[["three", "one"]]

,three,one
Ohio,2,0
Colorado,6,4
Utah,10,8
New York,14,12


In [69]:
data[:2] #slicing rows

,one,two,three,four
Ohio,0,1,2,3
Colorado,4,5,6,7


In [70]:
data[data["three"] > 5]

,one,two,three,four
Colorado,4,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


In [71]:
data < 5 #boolean DataFrame

,one,two,three,four
Ohio,True,True,True,True
Colorado,True,False,False,False
Utah,False,False,False,False
New York,False,False,False,False


In [72]:
data[data < 5] = 0
data

,one,two,three,four
Ohio,0,0,0,0
Colorado,0,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


#### selecting on DataFrame with loc and iloc

In [73]:
data
data.loc["Colorado"]
data.loc[["Colorado", "New York"]]

,one,two,three,four
Colorado,0,5,6,7
New York,12,13,14,15


In [74]:
data.loc["Colorado", ["two", "three"]]

two      5
three    6
Name: Colorado, dtype: int64

In [75]:
data.iloc[2] #selecting the third row of the DataFrame
data.iloc[[2, 1]]
data.iloc[2, [3, 0, 1]] #selecting third row, and columns four, one, and two (in that order)
data.iloc[[1, 2], [3, 0, 1]] #selecting rows two and three, and columns four, one, and two (in that order)

,four,one,two
Colorado,7,0,5
Utah,11,8,9


In [76]:
data.loc[:"Utah", "two"]

Ohio        0
Colorado    5
Utah        9
Name: two, dtype: int64

In [77]:
data.iloc[:, :3][data.three > 5] # selecting all rows, and columns up to three, then filtering rows where column three is greater than 5

,one,two,three
Colorado,0,5,6
Utah,8,9,10
New York,12,13,14


Boolean arrays can be used only with loc

In [78]:
data.loc[data.three >=2]

,one,two,three,four
Colorado,0,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


In [79]:
ser = pd.Series(np.arange(3.))
ser

0    0.0
1    1.0
2    2.0
dtype: float64

In [80]:
ser2 = pd.Series(np.arange(3.), index=['a', 'b', 'c'])
ser2[-1] #negative indexing doesn't work with label based indexing, but works with position based indexing like iloc

/var/folders/9p/c9c7v_v53bd91dc97zsmsm740000gn/T/ipykernel_30191/3700578937.py:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  ser2[-1] #negative indexing doesn't work with label based indexing, but works with position based indexing like iloc


2.0

In [81]:
ser.iloc[-1]
ser[:2]

0    0.0
1    1.0
dtype: float64

### Pitfalls with chained indexing

In [82]:
data.loc[:, "one"] = 1
data

,one,two,three,four
Ohio,1,0,0,0
Colorado,1,5,6,7
Utah,1,9,10,11
New York,1,13,14,15


In [83]:
data.iloc[2] = 5
data

,one,two,three,four
Ohio,1,0,0,0
Colorado,1,5,6,7
Utah,5,5,5,5
New York,1,13,14,15


In [84]:
data.loc[data["four"] > 5] = 3
data

,one,two,three,four
Ohio,1,0,0,0
Colorado,3,3,3,3
Utah,5,5,5,5
New York,3,3,3,3


In [85]:
data.loc[data.three == 5, "three"] = 6
data

,one,two,three,four
Ohio,1,0,0,0
Colorado,3,3,3,3
Utah,5,5,6,5
New York,3,3,3,3


### Arithmetic and Data Alignment

In [86]:
s1 = pd.Series([7.3, -2.5, 3.4, 1.5], index=["a", "c", "d", "e"])
s2 = pd.Series([-2.1, 3.6, -1.5, 4, 3.1], 
                   index=["a", "c", "e", "f", "g"])
s1
s2

a   -2.1
c    3.6
e   -1.5
f    4.0
g    3.1
dtype: float64

In [87]:
s1 + s2 

a    5.2
c    1.1
d    NaN
e    0.0
f    NaN
g    NaN
dtype: float64

In [88]:
df1 = pd.DataFrame(np.arange(9.).reshape((3, 3)), columns=list("bcd"),
                     index=["Ohio", "Texas", "Colorado"])

df2 = pd.DataFrame(np.arange(12.).reshape((4, 3)), columns=list("bde"),
                        index=["Utah", "Ohio", "Texas", "Oregon"])

df1

df2

,b,d,e
Utah,0.0,1.0,2.0
Ohio,3.0,4.0,5.0
Texas,6.0,7.0,8.0
Oregon,9.0,10.0,11.0


In [89]:
df1 = pd.DataFrame({"A": [1, 2]})
df2 = pd.DataFrame({"B": [3, 4]})
df1
df2 
df1 + df2

,A,B
0,NaN,NaN
1,NaN,NaN


##### Arithmetic methods with fill values

In [90]:
df1 = pd.DataFrame(np.arange(12.).reshape((3, 4)), 
                   columns=list("abcd"))
df2 = pd.DataFrame(np.arange(20.).reshape((4, 5)),
                     columns=list("abcde"))
df2.loc[1, "b"] = np.nan
df1
df2
df1 + df2

,a,b,c,d,e
0,0.0,2.0,4.0,6.0,NaN
1,9.0,NaN,13.0,15.0,NaN
2,18.0,20.0,22.0,24.0,NaN
3,NaN,NaN,NaN,NaN,NaN


In [91]:
df1.add(df2, fill_value=0)

,a,b,c,d,e
0,0.0,2.0,4.0,6.0,4.0
1,9.0,5.0,13.0,15.0,9.0
2,18.0,20.0,22.0,24.0,14.0
3,15.0,16.0,17.0,18.0,19.0


In [92]:
1 / df1

,a,b,c,d
0,inf,1.000000,0.500000,0.333333
1,0.250,0.200000,0.166667,0.142857
2,0.125,0.111111,0.100000,0.090909


In [93]:
df1.rdiv(1) #same as 1/df1, but with reverse division so that the fill value of 0 is replaced with 1/0 = inf instead of 0/1 = 0

,a,b,c,d
0,inf,1.000000,0.500000,0.333333
1,0.250,0.200000,0.166667,0.142857
2,0.125,0.111111,0.100000,0.090909


In [94]:
df1.reindex(columns=df2.columns, fill_value=0)

,a,b,c,d,e
0,0.0,1.0,2.0,3.0,0
1,4.0,5.0,6.0,7.0,0
2,8.0,9.0,10.0,11.0,0


#### Operations between DataFrame and Series

In [95]:
arr = np.arange(12.).reshape((3, 4))
arr

array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]])

In [96]:
arr[0]

array([0., 1., 2., 3.])

In [97]:
arr - arr[0] #subtracting the first row of the array from all rows, broadcasting the subtraction across the array

array([[0., 0., 0., 0.],
       [4., 4., 4., 4.],
       [8., 8., 8., 8.]])

In [98]:
frame = pd.DataFrame(np.arange(12.).reshape((4, 3)),
                        columns=list("bde"),
                        index=["Utah", "Ohio", "Texas", "Oregon"])
series = frame.iloc[0]
frame

,b,d,e
Utah,0.0,1.0,2.0
Ohio,3.0,4.0,5.0
Texas,6.0,7.0,8.0
Oregon,9.0,10.0,11.0


In [99]:
series #since iloc was set to 0, displays the first row

b    0.0
d    1.0
e    2.0
Name: Utah, dtype: float64

In [100]:
frame - series

,b,d,e
Utah,0.0,0.0,0.0
Ohio,3.0,3.0,3.0
Texas,6.0,6.0,6.0
Oregon,9.0,9.0,9.0


In [101]:
series2 = pd.Series(np.arange(3), index=["b", "e", "f"])
series2

b    0
e    1
f    2
dtype: int64

In [102]:
frame + series2

,b,d,e,f
Utah,0.0,NaN,3.0,NaN
Ohio,3.0,NaN,6.0,NaN
Texas,6.0,NaN,9.0,NaN
Oregon,9.0,NaN,12.0,NaN


In [103]:
series3 = frame["d"]
frame

,b,d,e
Utah,0.0,1.0,2.0
Ohio,3.0,4.0,5.0
Texas,6.0,7.0,8.0
Oregon,9.0,10.0,11.0


In [104]:
series3

Utah       1.0
Ohio       4.0
Texas      7.0
Oregon    10.0
Name: d, dtype: float64

In [105]:
frame.sub(series3, axis="index")

,b,d,e
Utah,-1.0,0.0,1.0
Ohio,-1.0,0.0,1.0
Texas,-1.0,0.0,1.0
Oregon,-1.0,0.0,1.0


#### Function Application and Mapping

In [106]:
frame = pd.DataFrame(np.random.standard_normal((4, 3)),
                    columns=list("bde"),
                    index=["Utah", "Ohio", "Texas", "Oregon"])
frame

,b,d,e
Utah,0.459909,0.979912,-0.942680
Ohio,-2.424392,0.640376,-0.397212
Texas,0.498319,0.961702,-0.207245
Oregon,-0.784850,0.200475,0.428332


In [107]:
np.abs(frame) #absolute value

,b,d,e
Utah,0.459909,0.979912,0.942680
Ohio,2.424392,0.640376,0.397212
Texas,0.498319,0.961702,0.207245
Oregon,0.784850,0.200475,0.428332


In [108]:
def f1(x):
    return x.max() - x.min()
frame.apply(f1) 

b    2.922711
d    0.779436
e    1.371012
dtype: float64

In [109]:
frame.apply(f1, axis="columns")

Utah      1.922592
Ohio      3.064768
Texas     1.168947
Oregon    1.213182
dtype: float64

In [110]:
def f2(x):
    return pd.Series([x.min(), x.max()], index=["min", "max"])
frame.apply(f2)

,b,d,e
min,-2.424392,0.200475,-0.942680
max,0.498319,0.979912,0.428332


In [111]:
def my_format(x):
    return f"{x:.2f}" 
frame.applymap(my_format)

/var/folders/9p/c9c7v_v53bd91dc97zsmsm740000gn/T/ipykernel_30191/1369377372.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  frame.applymap(my_format)


,b,d,e
Utah,0.46,0.98,-0.94
Ohio,-2.42,0.64,-0.40
Texas,0.50,0.96,-0.21
Oregon,-0.78,0.20,0.43


In [112]:
frame["e"].map(my_format) #applying a function to each element.value

Utah      -0.94
Ohio      -0.40
Texas     -0.21
Oregon     0.43
Name: e, dtype: object

### Sorting and Ranking

In [113]:
obj = pd.Series(np.arange(4), index=["d", "a", "b", "c"])
obj

d    0
a    1
b    2
c    3
dtype: int64

In [114]:
obj.sort_index() #sort by index

a    1
b    2
c    3
d    0
dtype: int64

In [115]:
frame = pd.DataFrame(np.arange(8).reshape((2, 4)),
                     index=["three", "one"],
                    columns=["d", "a", "b", "c"])
frame

,d,a,b,c
three,0,1,2,3
one,4,5,6,7


In [116]:
frame.sort_index() #sort by index, default is axis=0, so sorting by rows

,d,a,b,c
one,4,5,6,7
three,0,1,2,3


In [117]:
frame.sort_index(axis="columns")

,a,b,c,d
three,1,2,3,0
one,5,6,7,4


In [118]:
frame.sort_index(axis="columns", ascending=False) #default is ascending=True, so setting to False sorts in descending order

,d,c,b,a
three,0,3,2,1
one,4,7,6,5


In [119]:
obj = pd.Series([4, 7, -3, 2])
obj.sort_values()

2   -3
3    2
0    4
1    7
dtype: int64

In [120]:
obj = pd.Series([4, np.nan, 7, np.nan, -3, 2])
obj.sort_values()

4   -3.0
5    2.0
0    4.0
2    7.0
1    NaN
3    NaN
dtype: float64

In [121]:
obj.sort_values(na_position="first") #putting NaN values first instead of last

1    NaN
3    NaN
4   -3.0
5    2.0
0    4.0
2    7.0
dtype: float64

In [122]:
frame = pd.DataFrame({"b": [4, 7, -3, 2], "a": [0, 1, 0, 1]})
frame

,b,a
0,4,0
1,7,1
2,-3,0
3,2,1


In [123]:
frame.sort_values("b") #sort by values in column b

,b,a
2,-3,0
3,2,1
0,4,0
1,7,1


In [124]:
frame.sort_values(["a", "b"])

,b,a
2,-3,0
0,4,0
3,2,1
1,7,1


In [125]:
obj = pd.Series([7, -5, 7, 4, 2, 0, 4])
obj.rank()

0    6.5
1    1.0
2    6.5
3    4.5
4    3.0
5    2.0
6    4.5
dtype: float64

pd.Series.rank uses averages. mean of tied ranks

In [126]:
obj.rank(method="first")

0    6.0
1    1.0
2    7.0
3    4.0
4    3.0
5    2.0
6    5.0
dtype: float64

In [127]:
obj.rank(ascending=False)

0    1.5
1    7.0
2    1.5
3    3.5
4    5.0
5    6.0
6    3.5
dtype: float64

### Summarizing and Computing Descriptive Statistics

In [128]:
df = pd.DataFrame([[1.4, np.nan], [7.1, -4.5],
                   [np.nan, np.nan], [0.75, -1.3]],
                    index=["a", "b", "c", "d"],
                    columns=["one", "two"])
df

,one,two
a,1.40,NaN
b,7.10,-4.5
c,NaN,NaN
d,0.75,-1.3


In [129]:
df.sum() #sum of each column, excluding NaN values by default
df.sum(axis="columns") #sum of each row, excluding NaN values by default

a    1.40
b    2.60
c    0.00
d   -0.55
dtype: float64

In [130]:
df.sum(axis="index", skipna=False) #sum of each column, including NaN values (any NaN values will make the sum NaN) 

one   NaN
two   NaN
dtype: float64

In [131]:
df.mean(axis="columns")
df.idxmax()

one    b
two    d
dtype: object

#### Unique values, value counts, and membership

In [132]:
obj = pd.Series(["c", "a", "d", "a", "a", "b", "b", "c", "c"])
uniques = obj.unique()
uniques
obj.value_counts()

c    3
a    3
b    2
d    1
Name: count, dtype: int64

In [133]:
pd.value_counts(obj.to_numpy(), sort=False)

/var/folders/9p/c9c7v_v53bd91dc97zsmsm740000gn/T/ipykernel_30191/164454357.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  pd.value_counts(obj.to_numpy(), sort=False)


c    3
a    3
d    1
b    2
Name: count, dtype: int64

---
### Exercises — Ch 5

**Build and index a vitals DataFrame**

Construct a DataFrame with columns `['patient_id', 'hr', 'spo2', 'temp']` and 10 rows of made-up values.
Use `.loc` to select rows where `hr > 100`. Use `.iloc` to select the first 3 rows and first 2 columns.
Confirm the results are what you expected.

In [134]:
data = {'patient_id': [np.random.randint(1000) for _ in range(10)],
        'hr': np.random.randint(60, 100, size=10), 
        'spo2': np.random.randint(90, 100, size=10),
        'temp': np.random.uniform(36.5, 37.5, size=10)}
frame = pd.DataFrame(data)
frame

,patient_id,hr,spo2,temp
0,916,70,90,37.320714
1,611,82,97,36.824959
2,183,82,99,37.369198
3,785,85,91,36.721185
4,176,79,96,36.935986
5,390,99,98,37.358134
6,279,94,97,37.392339
7,214,71,94,36.595219
8,85,80,92,37.169100
9,652,87,93,37.387224


In [135]:
frame.loc[frame['hr'] > 100]

,patient_id,hr,spo2,temp


In [136]:
frame.iloc[:3, :2]

,patient_id,hr
0,916,70
1,611,82
2,183,82


**Index alignment**

Make two Series:
```
a = pd.Series([1, 2, 3], index=['a','b','c'])
b = pd.Series([10, 20], index=['b','c'])
```
Add them. Predict what happens to index `'a'` before running.

In [137]:
a = pd.Series([1, 2, 3], index=["a", "b", "c"])
b = pd.Series([10, 20], index=['b','c'])
a

a    1
b    2
c    3
dtype: int64

In [138]:
a + b #since b doesn't have row b, a comes up as NaN. rest of the rows are added together.

a     NaN
b    12.0
c    23.0
dtype: float64

---
## Chapter 6 — Data Loading, Storage, and File Formats (skim)

skim because this is just file formatting and basic commands. better to practice w/ real data
- "read_" functions load data into pandas. converting text data into a DataFrame. falls into:
    - indexing
    - type inference and data conversion
    - date and time parsing
    - iterating
    - unclean data issues 

---
## Chapter 7 — Data Cleaning

### Missing data — dropna / fillna / isnull

### Outlier detection and clip

In [139]:
# type code as you read — predict output before running

### pd.cut / pd.qcut

In [140]:
# type code as you read — predict output before running

### apply / map

In [141]:
# type code as you read — predict output before running

### pd.get_dummies

In [142]:
# type code as you read — predict output before running

---
### Exercises — Ch 7

**Missing vitals**

Build a DataFrame with columns `['hr', 'spo2', 'temp']` and 20 rows.
Randomly set ~15% of values to NaN using `rng.choice`.
Count missing values per column. Then fill `hr`/`spo2` with forward-fill and `temp` with the column mean.
Verify no NaNs remain.

**Outlier clip**

Given `spo2 = pd.Series([98, 102, 75, 99, -3, 95, 101, 88])`,
identify values outside physiologic range (70–100) using a boolean mask.
Replace them with NaN, then forward-fill.

**Binning HR into clinical labels**

Given `hr = pd.Series([55, 78, 105, 160, 190, 95, 130])`,
use `pd.cut` with `bins=[0, 60, 100, 160, 300]` and `labels=['brady','low-normal','normal','tachy']`.

**Dummy variables**

Given a DataFrame with a `'sex'` column containing `'M'`/`'F'` and a `'ga_group'` column with `'preterm'`/`'term'`,
use `pd.get_dummies` with `drop_first=True` and confirm the output shape.